File: python/golden_board/recipe_wire_v1.py
"""Bounded compact recipe framing with unchanged logical VM validation.

Diagnostic profiles 1..8 are explicit to this codec. Historical public v0
decoding/evaluation retain version zero and profiles 1..7; profile 8 is not
promoted by encoding or evaluating its recipe package.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

from . import bootstrap


__all__ = (
    "RecipePackageV1",
    "encode_recipe_package_v1",
    "expand_recipe_package_v1",
    "decode_recipe_package_v1",
    "evaluate_recipe_v1",
)


@dataclass(frozen=True, slots=True)
class RecipePackageV1:
    profile_version: int
    encoded: bytes
    logical: bootstrap.RecipePackage


def _uint(raw: bytes, offset: int, width: int) -> int:
    return int.from_bytes(raw[offset:offset + width], "big")


def _reject(path: str) -> None:
    raise bootstrap.BootstrapReject(bootstrap.RECIPE, path)


def _arity(opcode: int) -> int:
    if not 1 <= opcode <= 25:
        _reject("compact.node.opcode")
    if opcode in (1, 2, 24, 25):
        return 0
    if opcode in (5, 14, 22):
        return 1
    if opcode in (3, 20, 23):
        return 3
    return 2


def _fields(opcode: int) -> tuple[int, bool, bool]:
    return _arity(opcode), opcode in (2, 5, 22), opcode in (1, 5, 14, 22, 25)


def encode_recipe_package_v1(expanded_raw: bytes, expected_profile: int) -> bytes:
    """Validate a logical package before omitting its uniquely implied fields."""

    logical = bootstrap._decode_recipe_package(
        expanded_raw, expected_profile, maximum_profile_version=8
    )
    table_end = 64 + sum(16 + len(table.payload) for table in logical.tables)
    encoded = bytearray(expanded_raw[:table_end])
    encoded[8:10] = b"\0\1"
    offset = table_end
    for recipe in logical.recipes:
        node_start = offset + 32 + 12 * (len(recipe.inputs) + len(recipe.outputs))
        header = bytearray(expanded_raw[offset:node_start])
        body = bytearray()
        for index in range(len(recipe.nodes)):
            node = expanded_raw[node_start + 32 * index:node_start + 32 * (index + 1)]
            arity, auxiliary, immediate = _fields(node[2])
            body.extend(node[2:8])
            body.extend(node[10:10 + 2 * arity])
            if auxiliary:
                body.extend(node[18:20])
            if immediate:
                body.extend(node[24:32])
        header[28:32] = (len(header) + len(body)).to_bytes(4, "big")
        encoded.extend(header)
        encoded.extend(body)
        offset += recipe.recipe_bytes
    encoded[32:36] = len(encoded).to_bytes(4, "big")
    return bytes(encoded)


def decode_recipe_package_v1(compact_raw: bytes, expected_profile: int) -> RecipePackageV1:
    """Bound wire framing before expansion, then apply every logical v0 check."""

    maximum = bootstrap.RECIPE_PACKAGE_MAX
    if type(compact_raw) is not bytes or not 64 <= len(compact_raw) <= maximum:
        _reject("compact.length")
    if (
        compact_raw[:8] != b"GBRECP0\0"
        or _uint(compact_raw, 8, 2) != 1
        or _uint(compact_raw, 10, 2) != 0
        or type(expected_profile) is not int
        or not 1 <= expected_profile <= 8
        or _uint(compact_raw, 12, 2) != expected_profile
        or _uint(compact_raw, 14, 2) != 0
        or any(compact_raw[48:64])
    ):
        _reject("compact.header")
    recipe_count = _uint(compact_raw, 16, 2)
    table_count = _uint(compact_raw, 18, 2)
    total_nodes = _uint(compact_raw, 20, 4)
    if (
        not 1 <= recipe_count <= 256
        or table_count > 4096
        or not 1 <= total_nodes <= bootstrap.RECIPE_NODE_MAX
        or _uint(compact_raw, 24, 4) > bootstrap.RECIPE_EDGE_MAX
        or _uint(compact_raw, 28, 4) > maximum
        or _uint(compact_raw, 32, 4) != len(compact_raw)
        or _uint(compact_raw, 36, 8) > bootstrap.RECIPE_STEP_MAX
        or _uint(compact_raw, 44, 4) > bootstrap.RECIPE_SCRATCH_MAX
    ):
        _reject("compact.ranges")
    # First validate framing and derive the complete expanded bound, without
    # allocating output. Table typing/scalars remain the logical parser's job.
    offset = 64
    for _ in range(table_count):
        if offset + 16 > len(compact_raw):
            _reject("compact.table.header")
        count = _uint(compact_raw, offset + 8, 4)
        payload = _uint(compact_raw, offset + 12, 4)
        if not 1 <= count <= maximum or payload > maximum:
            _reject("compact.table.ranges")
        end = offset + 16 + payload
        if end > len(compact_raw):
            _reject("compact.table.boundary")
        offset = end

    table_end = offset
    expanded_size = table_end
    node_sum = 0
    recipes = []
    for _ in range(recipe_count):
        if offset + 32 > len(compact_raw):
            _reject("compact.recipe.header")
        inputs = _uint(compact_raw, offset + 4, 2)
        outputs = _uint(compact_raw, offset + 6, 2)
        nodes = _uint(compact_raw, offset + 8, 4)
        if (
            inputs > 64
            or not 1 <= outputs <= 64
            or not 1 <= nodes <= bootstrap.RECIPE_NODE_MAX
            or inputs + nodes > 0xffff
            or _uint(compact_raw, offset + 12, 4) > bootstrap.RECIPE_EDGE_MAX
            or _uint(compact_raw, offset + 16, 8) > bootstrap.RECIPE_STEP_MAX
            or _uint(compact_raw, offset + 24, 4) > bootstrap.RECIPE_SCRATCH_MAX
        ):
            _reject("compact.recipe.ranges")
        node_sum += nodes
        if node_sum > total_nodes:
            _reject("compact.total_node_count")
        descriptor_bytes = 12 * (inputs + outputs)
        recipe_bytes = _uint(compact_raw, offset + 28, 4)
        if not 32 + descriptor_bytes + 6 * nodes <= recipe_bytes <= 32 + descriptor_bytes + 18 * nodes:
            _reject("compact.recipe.length")
        end = offset + recipe_bytes
        if end > len(compact_raw):
            _reject("compact.recipe.boundary")
        node_start = offset + 32 + descriptor_bytes
        expanded_size += 32 + descriptor_bytes + 32 * nodes
        if expanded_size > maximum:
            _reject("compact.expanded_length")
        node_offset = node_start
        for _ in range(nodes):
            if node_offset + 6 > end:
                _reject("compact.node.prefix")
            arity, auxiliary, immediate = _fields(compact_raw[node_offset])
            node_bytes = 6 + 2 * arity + 2 * auxiliary + 8 * immediate
            if node_offset + node_bytes > end:
                _reject("compact.node.boundary")
            node_offset += node_bytes
        if node_offset != end:
            _reject("compact.recipe.trailing_fields")
        recipes.append((offset, node_start, nodes))
        offset = end
    if node_sum != total_nodes or offset != len(compact_raw):
        _reject("compact.trailing_or_counts")

    # Every copy extent and implicit zero field is now bounded. Original bytes
    # are immutable, so this pass uses exactly the framing validated above.
    expanded = bytearray(expanded_size)
    expanded[:table_end] = compact_raw[:table_end]
    expanded[8:10] = b"\0\0"
    expanded[32:36] = expanded_size.to_bytes(4, "big")
    output = table_end
    for start, node_start, nodes in recipes:
        header_bytes = node_start - start
        expanded[output:output + header_bytes] = compact_raw[start:node_start]
        expanded[output + 28:output + 32] = (header_bytes + 32 * nodes).to_bytes(4, "big")
        output += header_bytes
        node_offset = node_start
        for index in range(nodes):
            arity, auxiliary, immediate = _fields(compact_raw[node_offset])
            expanded[output:output + 2] = (index + 1).to_bytes(2, "big")
            expanded[output + 2:output + 8] = compact_raw[node_offset:node_offset + 6]
            expanded[output + 8:output + 10] = arity.to_bytes(2, "big")
            node_offset += 6
            expanded[output + 10:output + 10 + 2 * arity] = compact_raw[node_offset:node_offset + 2 * arity]
            node_offset += 2 * arity
            if auxiliary:
                expanded[output + 18:output + 20] = compact_raw[node_offset:node_offset + 2]
                node_offset += 2
            if immediate:
                expanded[output + 24:output + 32] = compact_raw[node_offset:node_offset + 8]
                node_offset += 8
            output += 32
    logical = bootstrap._decode_recipe_package(
        bytes(expanded), expected_profile, maximum_profile_version=8
    )
    return RecipePackageV1(expected_profile, compact_raw, logical)


def expand_recipe_package_v1(compact_raw: bytes, expected_profile: int) -> bytes:
    """Return exact expanded bytes only after complete logical validation."""

    return decode_recipe_package_v1(compact_raw, expected_profile).logical.encoded


def evaluate_recipe_v1(
    package: RecipePackageV1, recipe_id: int, inputs: Sequence[bytes]
) -> bootstrap.RecipeResult:
    """Reparse retained compact bytes; cached logical fields are not authority."""

    if not isinstance(package, RecipePackageV1):
        raise TypeError("evaluate_recipe_v1 requires RecipePackageV1")
    trusted = decode_recipe_package_v1(package.encoded, package.profile_version)
    return bootstrap._evaluate_validated_recipe(trusted.logical, recipe_id, inputs)

File: crates/gb-bootstrap/src/recipe_wire_v1.rs
//! Explicit compact recipe wire admission; the existing VM semantics are shared.
//!
//! Historical public parsers retain version-zero/profile-1..7 admission. This
//! module expands only bounded version-one packages and independently validates
//! their complete logical program before publishing or evaluating it.

use crate::recipe::{
    EDGE_MAX, NODE_MAX, PACKAGE_MAX, RECIPE_MAX, RecipeOutcome, RecipePackage, RecipeValue,
    SCRATCH_MAX, STEP_MAX, TABLE_MAX, decode_recipe_package_admitted,
    evaluate_serialized_validated_recipe, evaluate_validated_recipe,
};
use crate::{BootstrapError, RejectCode, Result};

/// Parsed diagnostic view. Execution always reparses `encoded`; mutable cached
/// logical fields cannot bypass admission or alter the executed program.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RecipePackageV1 {
    pub encoded: Vec<u8>,
    pub logical: RecipePackage,
}

fn reject() -> BootstrapError {
    BootstrapError {
        code: RejectCode::Recipe,
        offset: None,
    }
}

fn take(raw: &[u8], first: usize, length: usize) -> Result<&[u8]> {
    let end = first.checked_add(length).ok_or_else(reject)?;
    raw.get(first..end).ok_or_else(reject)
}

fn u16_at(raw: &[u8], offset: usize) -> Result<u16> {
    Ok(u16::from_be_bytes(
        take(raw, offset, 2)?.try_into().map_err(|_| reject())?,
    ))
}

fn u32_at(raw: &[u8], offset: usize) -> Result<u32> {
    Ok(u32::from_be_bytes(
        take(raw, offset, 4)?.try_into().map_err(|_| reject())?,
    ))
}

fn u64_at(raw: &[u8], offset: usize) -> Result<u64> {
    Ok(u64::from_be_bytes(
        take(raw, offset, 8)?.try_into().map_err(|_| reject())?,
    ))
}

fn length_at(raw: &[u8], offset: usize) -> Result<usize> {
    usize::try_from(u32_at(raw, offset)?).map_err(|_| reject())
}

fn put_length(raw: &mut [u8], offset: usize, value: usize) -> Result<()> {
    let value = u32::try_from(value).map_err(|_| reject())?;
    raw.get_mut(offset..offset.checked_add(4).ok_or_else(reject)?)
        .ok_or_else(reject)?
        .copy_from_slice(&value.to_be_bytes());
    Ok(())
}

fn arity(opcode: u8) -> Result<u16> {
    match opcode {
        1 | 2 | 24 | 25 => Ok(0),
        5 | 14 | 22 => Ok(1),
        3 | 20 | 23 => Ok(3),
        4 | 6..=13 | 15..=19 | 21 => Ok(2),
        _ => Err(reject()),
    }
}

#[derive(Clone, Copy)]
struct NodeFields {
    arguments: usize,
    auxiliary: bool,
    immediate: bool,
    bytes: usize,
}

fn node_fields(opcode: u8) -> Result<NodeFields> {
    let arguments = usize::from(arity(opcode)?);
    let auxiliary = matches!(opcode, 2 | 5 | 22);
    let immediate = matches!(opcode, 1 | 5 | 14 | 22 | 25);
    Ok(NodeFields {
        arguments,
        auxiliary,
        immediate,
        bytes: 6 + arguments * 2 + usize::from(auxiliary) * 2 + usize::from(immediate) * 8,
    })
}

struct RecipeFrame {
    first: usize,
    input_bytes: usize,
    front_bytes: usize,
    node_count: usize,
    output_bytes: usize,
}

struct WirePlan {
    tables_end: usize,
    output_bytes: usize,
    recipes: Vec<RecipeFrame>,
}

/// Bound the complete expansion from headers before allocating its buffer.
/// Recipe ranges also bound every subsequent variable-width node read.
fn plan(raw: &[u8], expected_profile: u16, expanding: bool) -> Result<WirePlan> {
    if !(64..=PACKAGE_MAX).contains(&raw.len())
        || !(1..=8).contains(&expected_profile)
        || take(raw, 0, 8)? != b"GBRECP0\0"
        || u16_at(raw, 8)? != u16::from(expanding)
        || u16_at(raw, 10)? != 0
        || u16_at(raw, 12)? != expected_profile
        || u16_at(raw, 14)? != 0
        || take(raw, 48, 16)?.iter().any(|byte| *byte != 0)
        || length_at(raw, 32)? != raw.len()
    {
        return Err(reject());
    }
    let recipe_count = usize::from(u16_at(raw, 16)?);
    let table_count = usize::from(u16_at(raw, 18)?);
    let node_count = length_at(raw, 20)?;
    if !(1..=RECIPE_MAX).contains(&recipe_count)
        || table_count > TABLE_MAX
        || !(1..=NODE_MAX).contains(&node_count)
        || u64::from(u32_at(raw, 24)?) > EDGE_MAX
        || length_at(raw, 28)? > PACKAGE_MAX
        || u64_at(raw, 36)? > STEP_MAX
        || u64::from(u32_at(raw, 44)?) > SCRATCH_MAX
    {
        return Err(reject());
    }
    let mut cursor = 64usize;
    let mut table_payload = 0usize;
    for _ in 0..table_count {
        let header = take(raw, cursor, 16)?;
        let count = length_at(header, 8)?;
        let payload = length_at(header, 12)?;
        if !(1..=PACKAGE_MAX).contains(&count) || payload > PACKAGE_MAX {
            return Err(reject());
        }
        let record_length = 16usize.checked_add(payload).ok_or_else(reject)?;
        take(raw, cursor, record_length)?;
        cursor = cursor.checked_add(record_length).ok_or_else(reject)?;
        table_payload = table_payload.checked_add(payload).ok_or_else(reject)?;
        if table_payload > PACKAGE_MAX {
            return Err(reject());
        }
    }
    if table_payload != length_at(raw, 28)? {
        return Err(reject());
    }
    let tables_end = cursor;
    let mut expanded_bytes = tables_end;
    let mut output_bytes = tables_end;
    let mut observed_nodes = 0usize;
    let mut recipes = Vec::with_capacity(recipe_count);
    for _ in 0..recipe_count {
        let header = take(raw, cursor, 32)?;
        let inputs = usize::from(u16_at(header, 4)?);
        let outputs = usize::from(u16_at(header, 6)?);
        let nodes = length_at(header, 8)?;
        if inputs > 64
            || !(1..=64).contains(&outputs)
            || !(1..=NODE_MAX).contains(&nodes)
            || inputs.checked_add(nodes).ok_or_else(reject)? > NODE_MAX
        {
            return Err(reject());
        }
        let descriptors = inputs
            .checked_add(outputs)
            .and_then(|count| count.checked_mul(12))
            .ok_or_else(reject)?;
        let front = 32usize.checked_add(descriptors).ok_or_else(reject)?;
        let input_length = length_at(header, 28)?;
        let expanded_length = nodes
            .checked_mul(32)
            .and_then(|bytes| bytes.checked_add(front))
            .ok_or_else(reject)?;
        expanded_bytes = expanded_bytes
            .checked_add(expanded_length)
            .ok_or_else(reject)?;
        observed_nodes = observed_nodes.checked_add(nodes).ok_or_else(reject)?;
        if expanded_bytes > PACKAGE_MAX || observed_nodes > node_count || input_length > PACKAGE_MAX
        {
            return Err(reject());
        }
        let record = take(raw, cursor, input_length)?;
        let encoded_nodes = input_length.checked_sub(front).ok_or_else(reject)?;
        let result_length = if expanding {
            if encoded_nodes < nodes.checked_mul(6).ok_or_else(reject)?
                || encoded_nodes > nodes.checked_mul(18).ok_or_else(reject)?
            {
                return Err(reject());
            }
            expanded_length
        } else {
            if input_length != expanded_length {
                return Err(reject());
            }
            let mut length = front;
            for node in record[front..].chunks_exact(32) {
                length = length
                    .checked_add(node_fields(node[2])?.bytes)
                    .ok_or_else(reject)?;
            }
            length
        };
        output_bytes = output_bytes.checked_add(result_length).ok_or_else(reject)?;
        if output_bytes > PACKAGE_MAX {
            return Err(reject());
        }
        recipes.push(RecipeFrame {
            first: cursor,
            input_bytes: input_length,
            front_bytes: front,
            node_count: nodes,
            output_bytes: result_length,
        });
        cursor = cursor.checked_add(input_length).ok_or_else(reject)?;
    }
    if cursor != raw.len() || observed_nodes != node_count {
        return Err(reject());
    }
    Ok(WirePlan {
        tables_end,
        output_bytes,
        recipes,
    })
}

/// Structurally transcode within the same frozen maximum. Semantic validation
/// is performed by each caller, never inferred from successful framing alone.
fn transcode(raw: &[u8], expected_profile: u16, expanding: bool) -> Result<Vec<u8>> {
    let plan = plan(raw, expected_profile, expanding)?;
    let mut output = Vec::with_capacity(plan.output_bytes);
    output.extend_from_slice(&raw[..plan.tables_end]);
    output[8..10].copy_from_slice(&u16::from(!expanding).to_be_bytes());
    put_length(&mut output, 32, plan.output_bytes)?;
    for frame in plan.recipes {
        let record = take(raw, frame.first, frame.input_bytes)?;
        let result_start = output.len();
        output.extend_from_slice(&record[..frame.front_bytes]);
        put_length(&mut output, result_start + 28, frame.output_bytes)?;
        let mut cursor = frame.front_bytes;
        for ordinal in 0..frame.node_count {
            if expanding {
                let opcode = *record.get(cursor).ok_or_else(reject)?;
                let fields = node_fields(opcode)?;
                let node = take(record, cursor, fields.bytes)?;
                let id = u16::try_from(ordinal + 1).map_err(|_| reject())?;
                let mut expanded = [0u8; 32];
                expanded[..2].copy_from_slice(&id.to_be_bytes());
                expanded[2..8].copy_from_slice(&node[..6]);
                expanded[8..10].copy_from_slice(&(fields.arguments as u16).to_be_bytes());
                let arguments_end = 6 + 2 * fields.arguments;
                expanded[10..10 + 2 * fields.arguments].copy_from_slice(&node[6..arguments_end]);
                let mut field_cursor = arguments_end;
                if fields.auxiliary {
                    expanded[18..20].copy_from_slice(&node[field_cursor..field_cursor + 2]);
                    field_cursor += 2;
                }
                if fields.immediate {
                    expanded[24..32].copy_from_slice(&node[field_cursor..field_cursor + 8]);
                }
                output.extend_from_slice(&expanded);
                cursor = cursor.checked_add(fields.bytes).ok_or_else(reject)?;
            } else {
                let node = take(record, cursor, 32)?;
                let fields = node_fields(node[2])?;
                output.extend_from_slice(&node[2..8]);
                output.extend_from_slice(&node[10..10 + 2 * fields.arguments]);
                if fields.auxiliary {
                    output.extend_from_slice(&node[18..20]);
                }
                if fields.immediate {
                    output.extend_from_slice(&node[24..32]);
                }
                cursor = cursor.checked_add(32).ok_or_else(reject)?;
            }
        }
        if cursor != record.len() || output.len() - result_start != frame.output_bytes {
            return Err(reject());
        }
    }
    if output.len() != plan.output_bytes {
        return Err(reject());
    }
    Ok(output)
}

/// Encode a fully validated expanded logical package; profile 8 is admitted only
/// by this explicit diagnostic entry point, never by the old public parser.
pub fn encode_recipe_package_v1(expanded: &[u8], expected_profile: u16) -> Result<Vec<u8>> {
    decode_recipe_package_admitted(expanded, expected_profile, 8)?;
    transcode(expanded, expected_profile, false)
}

fn expand_validated(raw: &[u8], expected_profile: u16) -> Result<(Vec<u8>, RecipePackage)> {
    let expanded = transcode(raw, expected_profile, true)?;
    let logical = decode_recipe_package_admitted(&expanded, expected_profile, 8)?;
    Ok((expanded, logical))
}

/// Recover exact expanded framing only after complete logical validation.
pub fn expand_recipe_package_v1(raw: &[u8], expected_profile: u16) -> Result<Vec<u8>> {
    Ok(expand_validated(raw, expected_profile)?.0)
}

/// Decode and validate both compact framing and the complete unchanged VM rules.
pub fn decode_recipe_package_v1(raw: &[u8], expected_profile: u16) -> Result<RecipePackageV1> {
    let (_, logical) = expand_validated(raw, expected_profile)?;
    Ok(RecipePackageV1 {
        encoded: raw.to_vec(),
        logical,
    })
}

/// Evaluate newly parsed bytes, not the cached diagnostic logical fields.
pub fn evaluate_recipe_v1(
    package: &RecipePackageV1,
    recipe_id: u16,
    inputs: &[RecipeValue],
) -> Result<RecipeOutcome> {
    let (_, logical) = expand_validated(&package.encoded, package.logical.profile_version)?;
    evaluate_validated_recipe(&logical, recipe_id, inputs)
}

/// Evaluate a serialized interface with exact status/output suppression.
pub fn evaluate_serialized_recipe_v1(
    package: &RecipePackageV1,
    recipe_id: u16,
    input: &[u8],
) -> Result<Vec<u8>> {
    let (_, logical) = expand_validated(&package.encoded, package.logical.profile_version)?;
    evaluate_serialized_validated_recipe(&logical, recipe_id, input)
}
